dev/core#914 rewrite submitOnce function

colemanw opened the pull request:
Overview
----------------------------------------
Fix https://lab.civicrm.org/dev/core/issues/914

Before
----------------------------------------
SubmitOnce applied on a per-button basis. Doesn't work for forms with multiple buttons. Changes button text.

After
----------------------------------------
SubmitOnce applied on a per-form basis. Works with multiple buttons. Doesn't change button text but adds a spinner icon to the button while submitting.

demeritcowboy commented:
As discussed, while addressing the enter key submitting the form on firefox is out of scope, since this does prevent double-hitting the enter key in the standalone form, note it doesn't prevent double-hitting the enter key in the popup. If you double-hit the enter key on firefox in the popup you get duplicates.

eileenmcnaughton commented:
test this please

demeritcowboy commented:
Just cross-referencing https://lab.civicrm.org/dev/core/issues/943

demeritcowboy commented:
Checking multiple enter-key hits on chrome, you also get duplicates for the popup version. So in this PR it seems to work the same for chrome and firefox - double-hitting the enter key prevents duplicates for standalone but not for popup.

colemanw commented:
@demeritcowboy this is ready for final review. I've fixed it to work with dialogs, preventing double-submit-on-enter and spinning the icon of the clicked button in the dialog.

I've deprecated the old `submitOnce` javascript fn and switched all forms to use the new function.

IMO the new function works so well that it ought to be enabled for all forms, but for now I haven't changed the number of forms using it (very few) and have left it opt-in rather than opt-out.

The contribution and event forms ought to get tested, as they do use submitOnce and I've switched them to the new fn.

demeritcowboy commented:
I have a couple other things to test still but I'm seeing the following:
1. Create a custom field for activities that's required.
2. Do the popup version of a meeting and don't fill out the custom field.
3. Click save.
4. You'll get the validation error. Now fill out the custom field.
5. You can't click save.
6. You can also get this without a custom field by doing something like clearing out the date field and then clicking save.

colemanw commented:
Thanks @demeritcowboy. I've added a few lines to make it play nice with jQuery validate. Should be better now.

demeritcowboy commented on CRM/Event/Form/Registration/AdditionalParticipant.php:
```diff
@@ -210,7 +210,7 @@ public function buildQuickForm() {
     $js = NULL;
     if ($this->isLastParticipant(TRUE) && empty($this->_values['event']['is_monetary'])) {
-      $js = ['onclick' => "return submitOnce(this,'" . $this->_name . "','" . ts('Processing') . "');"];
+      $this->submitOnce = TRUE;
```
I'm not sure that setting the value here in buildForm() actually works? It seems like you have to set it at the form level like in CRM/Event/Form/Registration/Confirm.php. In particular when I am at this form the `<form>` tag doesn't have the data-submit-once attribute set and running a debugger it doesn't trigger the submitOnce function. E.g. when I go through it for this event (as anonymous) http://core-14519-39tw3.test-1.civicrm.org:8001/civicrm/event/info?reset=1&id=7

seamuslee001 replied:
@demeritcowboy that additionalParticipant will only show if you are registering > 1 participant for the event and would only show on the 2nd and other n ones what the problem more likely is with the location of this https://github.com/civicrm/civicrm-core/pull/14519/files#diff-0e6b4e48225761b02cc269e604299d31R467 but it equally could apply here i'm not sure if doing the submit once on just non monetary events makes sense surely we would want it on all events right?

demeritcowboy replied:
Right for additionalParticipant it would only be on 2nd+, but it doesn't seem to get applied when you set the property in buildQuickForm. And yes it's the same issue for the other places where it's only set in buildQuickForm.

Regarding monetary vs nonmonetary the plan eventually I guess would to be to set it everywhere on all forms, just at the moment this PR is only touching some places.

demeritcowboy replied:
D'oh. I see it now. `is_monetary` means "Paid Event"=yes.

demeritcowboy commented:
@colemanw I'm done testing with Firefox. I'll do some more testing with Chrome but I think it's ready for someone else to take a run at it.

demeritcowboy commented:
Did some more testing with Chrome.

Also because of https://lab.civicrm.org/dev/core/issues/1045 and 1051 I ended up testing repeating activities.

Noticed that CRM/Upgrade/Form.php seems to be the only place that calls addDefaultButtons() with the submitOnce parameter TRUE, but I couldn't see where it was used. Tried running an upgrade but didn't see a step where it showed up.

So I think I have nothing left to test here. Everything seems to be working. And of the forms I accidentally tested where the new submitOnce _isn't_ used they seem to still work too.

----
Ref:
* Activities
* Mass SMS (on the final page)
* Contribution pages (with/without confirmation page enabled).
* Event registration and last of the additional participant pages (if no fees).
* New Case (save, save+new).
* Case Activity
* Variations:
  * Standalone
  * Popup
  * Mouse
  * Double-click
  * Enter Key
  * Double-hit enter key
  * Edit form
  * "Go Back"
  * Cancel.
* Delete in standalone mode (e.g. delete meeting) is the same activity form so also has submitOnce.

seamuslee001 commented:
I'm going to merge this based on @demeritcowboy 's testing which to me seems very extensive. This will also mean it hits the 5.16 RC as well which would give others more chance to test it.

eileenmcnaughton commented:
Huge thanks to @colemanw and @demeritcowboy for working hard on a solid fix for an ongoing issue

FYI @mattwire
